src/linker/Linker.Steps/ResolveFromAssemblyStep.cs

```diff
@@ -63,8 +63,13 @@ protected override void Process ()
 			if (_assembly != null)
 				Context.Resolver.CacheAssembly (_assembly);
 
-			AssemblyDefinition assembly = _assembly ?? Context.Resolve (_file);
+			var ignoreUnresolved = Context.Resolver.IgnoreUnresolved;
```

Comment:
Maybe add a comment why we're making this dance

Reply:
I don't think this is the right place to fix it. Please change `Assembly::ResolveFromReferences` instead of this hacking

Reply:
I don't really like the solution either, but I don't know if there's a better way. `AssemblyResolver::ResolveFromReferences` is called by all assembly resolution everywhere. The change needs to be targetted to the `ResolveFromAssemblyStep` - basically we still want to behave as per the setting (`IgnoreUnresolved`) everywhere, except when resolving assembly names which are specified directly on the command line, so `-r`. For those we would like to fail always, regardless of the setting. Making a change in the `AssemblyResovler` would change the behavior everywhere, not just for the command line inputs.

Maybe a better way to fix this would be to let let the `Context.Resovle` return `null` and check for that explicitly. If we want custom error message we would have to also try/catch `AssemblyResolutionException` as the `Resolve` may also throw in some cases.

Reply:
>except when resolving assembly names which are specified directly on the command line, so -r

I agree, which is what https://github.com/mono/linker/blob/6e0105e0fc16b959510480298599ca09a5509a21/src/linker/Linker/AssemblyResolver.cs#L114 does. The `_references` is populated only from `-r` settings. Moving that call into separate try/catch with different logging should resolve this issue better.

Reply:
As far as I can tell the `_references` is populated from `--reference` command line parameters:
https://github.com/mono/linker/blob/6e0105e0fc16b959510480298599ca09a5509a21/src/linker/Linker/Driver.cs#L290-L294

This explicitly mentions that it's different from `-r`. The `_references` contains file paths (or should contain file paths, otherwise it will break elsewhere). The `-r` specifies simple assembly name. Both should fail if they don't exist:
* for `--reference` we will eventually try to open it as a file path (`ModuleDefinition.ReadModule`)
* for `-r` we run it through the `Context.Resolve` (so it might be resolved from the `--references` or through probing)

Reply:
You are right, I keep forgetting about the `-r` hack.